scan-irb: Accept dynamic swab and staff codes

kairstenfay opened the pull request:
We now have SCAN IRB studies that have multiple swab types and multiple
~collection methods (self-collected or staff-collected)~ observation methods. Replace the
hard-coded values in the current, English template to account for this.

This depends on concurrent changes in the backoffice RoR workflow. ~For example, we need to converge on what these swab type and collection method columns will be called and what set of values they'll contain.~

We will also need translations for the new swab type and the "staff collected" collection method, so part of this PR is also to vet the language I used in these templates.

kairstenfay committed:
scan-irb: Accept dynamic swab and staff codes

We now have SCAN IRB studies that have multiple swab types and multiple
collection methods (self-collected or staff-collected). Replace the
hard-coded values in the current, English template to account for this.

kairstenfay commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Update lab result report template to accept templated text around swab type and staff vs. non-staff collection](https://trello.com/c/QGQAmt6E/678-update-lab-result-report-template-to-accept-templated-text-around-swab-type-and-staff-vs-non-staff-collection)

joverlee521 commented on scan-irb/report-en.tex:
```diff
@@ -58,7 +58,23 @@
 \item [\textbf{Date of Birth}] \VAR{birth_date|e}
-\item [\textbf{Specimen}] Midnasal, self-collected
+\item [\textbf{Specimen}]
+  %- if swab_type == "mts"
+  Midnasal,
+  %- elif swab_type == "ans"
+  Anterior,
+  %- else
+  \errmessage{Invalid swab_type "\VAR{swab_type}"}
+  %- endif
```
The current values in the specimen manifest dropdown menu are `mtb`, `ans`, `np`, and `unk`.

We should definitely error out for `unk` but we should include the other values here.

kairstenfay replied:
Thanks for this info!

joverlee521 commented on scan-irb/report-en.tex:
```diff
@@ -58,7 +58,23 @@
+  \errmessage{Invalid swab_type "\VAR{swab_type}"}
+  %- endif
+
+  %- if staff_collected == "1"
+  staff collected
+  %- elif staff_collected == "0"
+  self-collected
+  %- else
+  \errmessage{Invalid staff_collected "\VAR{staff_collected}"}
+  %- endif
```
FYI, UWGreek samples are still self-collected (see Izzy's [comment](https://seattle-flu-study.slack.com/archives/C017M3GJ35X/p1594940377055900?thread_ts=1594940274.054800&cid=C017M3GJ35X)). Not sure if this is a change we want to add now or later.

tsibley replied:
Yep. I believe the initial discussion with Izzy of changes needed was about self-vs-staff collection, but the final outcome was that the change is about staff-observed or not and how the specimen is transported back to the lab. Thus, I think we want a slightly different name than `staff_collected`.

kairstenfay replied:
Great, let's discuss in [this comment thread](https://github.com/seattleflu/id3c-customizations/pull/119#discussion_r458441591).

jccraft-uw commented on scan-irb/example-params.csv:
```diff
@@ -1,7 +1,7 @@
-FFFFFFFF,Positive Pat,2020-01-01,2020-06-10,2020-06-12,positive
+qrcode,pat_name,birth_date,collect_ts,result_ts,status_code,swab_type,staff_collected
+AAAAAAAA,Not-received Ned,2020-01-01,,,not-received,ans,0
+BBBBBBBB,Pending Patrick,2020-01-01,2020-06-10,,pending,mts,0
```
Was the idea that the view that produces the results would do the translating and the template wouldn't have to? That is, would we see "mtb" in the CSV file or "Midnasal"?

tsibley replied:
Good thing to consider. I think it could go either way. If the view does translate, then we need to make sure that unknown/null values from the view ends up causing those reports to fail. By translating in the template, we kinda get that assurance as a side-effect.

joverlee521 commented:
Reminder: We need to backfill `sample.details.swab_type` for all samples so that the lab-report generation doesn't error out:
```sql
begin;

update warehouse.sample set details = details || '{"swab_type": "mtb", "collection_matrix":"utm_vtm"}';
```

tsibley commented:
@joverlee521 Good reminder. That `update` may want to `coalesce(details, '{}') || …` so that null `details` column doesn't skip the update.

tsibley commented on scan-irb/report-en.tex:
```diff
@@ -58,7 +58,23 @@
+  %- elif swab_type == "ans"
+  Anterior,
+  %- else
+  \errmessage{Invalid swab_type "\VAR{swab_type}"}
```
TIL about `\errmessage`, nice! I clearly didn't know about it when I did my divide-by-zero hack elsewhere in this doc. I assume you tested and it works as expected?

kairstenfay replied:
Yes it does. Thank you TeX stack exchange for all of my LaTeX knowledge. 

Should I use `\errmessage` to replace the divide-by-zero code you wrote?

tsibley replied:
> Should I use `\errmessage` to replace the divide-by-zero code you wrote?

Sure!

kairstenfay replied:
I'll do this.. eventually!

tsibley reviewed (changes requested):
This is a great first pass, thanks! I believe the associated Trello card has other necessary language changes too.

jccraft-uw commented on scan-irb/report-en.tex:
```diff
@@ -58,7 +58,23 @@
+  %- elif swab_type == "ans"
+  Anterior,
+  %- else
+  \errmessage{Invalid swab_type "\VAR{swab_type}"}
```
Do you need to accommodate swab type "none"? See Slack conversation https://seattle-flu-study.slack.com/archives/CCAEWSFTK/p1595442310319000

kairstenfay replied:
We would filter out swab types of `none`, so we should throw an error if they slip in here.

kairstenfay replied:
Oh shoot, I conflated this with `unk` in my head. I don't know if we know how to handle `none` right now. We don't have the language around it.

jccraft-uw replied:
See the latest in the Slack thread. It sounds like the "none" swab-type case are samples/results that we won't generate PDFs for.

kairstenfay commented:
I implemented the requested changes and normalized on `staff_observed` true/false

kschwabefry commented on scan-irb/report-en.tex:
```diff
@@ -105,11 +117,21 @@
+
+  %- elif staff_observed == "false"
+\item
+  You collected this sample by yourself (or for someone else) and it was labeled
```
Does this new text block correlate to the testing sweeps from this week and last week? Would those samples collected from the sweeps have staff_observed == true or false?

kairstenfay replied:
Yes, this is in preparation for the kiosk (in-person) testing and the Greek sweeps. Those encounters will have staff_observed == 'True'

Writing this out made me reconsider that we may want to use python native boolean values `True`/`False` instead of lowercase "true"/"false" (although LaTeX will likely have to treat the values all as strings)

kairstenfay replied:
Oh, yes you're right. I accidentally got these flipped backwards. Thank you for catching that!

kairstenfay commented:
I just force-pushed a fixup commit thanks to @kschwabefry catching that I flipped the conditionals around after renaming them!

jccraft-uw commented on scan-irb/report-en.tex:
```diff
@@ -260,11 +281,21 @@
 
 \item
-
+  %- if staff_observed == "false"
```
Is this not repeating what starts on line 121? I've stared at that a couple of times and can't tell the difference. Sorry if I'm reading it wrong?

kairstenfay replied:
Yes, this paragraph is duplicated in the template on master.

jccraft-uw replied:
Ahhhh.

tsibley committed:
scan-irb: Error on unknown staff_observed values too

Otherwise we might include neither paragraph.

tsibley commented:
Looks good. I pushed a commit to make the `staff_observed` bit more defensive: 
186a015.

kairstenfay commented:
> Looks good. I pushed a commit to make the `staff_observed` bit more defensive:
> [186a015](https://github.com/seattleflu/lab-result-reports/commit/186a0153e1c5460a7208f35d59ba340d2e80c586).

Oh, thanks. I forgot that I was no longer checking for unknown `staff_observed` codes at the top (because we're still only doing "self-collected" swabs).

Are we good to merge?

joverlee521 commented:
> Are we good to merge?

We should probably wait to merge & deploy after all the upstream work is deployed

tsibley commented:
Ahh, the `staff_observed` values will be `t/f` unless I convert them to `true/false` in the PII join.

kairstenfay commented:
> Ahh, the `staff_observed` values will be `t/f` unless I convert them to `true/false` in the PII join.

`t/f` is fine with me especially if it's simpler on your side.

tsibley committed:
scan-irb: staff_collected uses t/f instead of true/false

tsibley commented:
> t/f is fine with me especially if it's simpler on your side.

Pushed a commit adjusting this.

tsibley committed:
scan-irb: Remove support for NP swabs

We don't plan to ever use NP swabs for participants since they can't be
self-collected, so we won't ever be returning results for them.  The
lab's swab type field includes them, however, because occasionally the
lab does test NP swabs from other sources.  By removing the conditional
clause in the template, we'll produce an error if one slips through
somehow.

tsibley committed:
scan-irb: Spanish template support for swab_type and staff_observed

tsibley commented:
Dropped support for NP swabs, since we should never be returning NP results, and added swab_type/staff_observed support for Spanish template.

tsibley commented:
One final item:

- [ ] Bump version of lab-result-reports image once {available, known, https://github.com/seattleflu/backoffice/pull/65 is merged}.

tsibley committed:
scan-irb: Update Spanish translation for MTB swab per Misja

Makes it consistent with the new (more precise) translation for anterior
nares swab.

kairstenfay committed:
scan-irb: Traditional Chinese template support for swab, staff variables

kairstenfay commented:
I just added traditional Chinese support.

kairstenfay commented:
Chinese and Spanish have been validated.

tsibley commented:
@kairstenfay Thanks! I [amended your commit](https://github.com/seattleflu/lab-result-reports/compare/81d94ad1c0da62eb0f56558d2b87c00b1b27c049..74819118a4f81384b59a13c2b2daae93442ebc66) to include the final `%- else` for unknown `staff_observed` values.

kairstenfay commented:
> @kairstenfay Thanks! I [amended your commit](https://github.com/seattleflu/lab-result-reports/compare/81d94ad1c0da62eb0f56558d2b87c00b1b27c049..74819118a4f81384b59a13c2b2daae93442ebc66) to include the final `%- else` for unknown `staff_observed` values.

Ugh, everything just starts to look the same after awhile... thank you.
